{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "mexican-fruit",
   "metadata": {},
   "source": [
    "# Finding CRISPR arrays\n",
    "## Specification\n",
    "## Program Name: randomizedMotifSearch.py\n",
    "\n",
    "## Input/Output: STDIN, STDOUT\n",
    "\n",
    "## Options:\n",
    "\n",
    "- -i iterations (int)\n",
    "- -k motif length (int)\n",
    "- -p pseudocount (float)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "happy-suicide",
   "metadata": {},
   "outputs": [],
   "source": [
    "Ex: python randomizedMotifSearch.py -i 1000 -k 13 -p 1 < input.fa > output.out\n",
    "\n",
    "Inspection Notebook example:\n",
    "main ( \"input.fa\", options = [ \"-i 1000\", \"-k 13\", \"-p 1\"] )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "particular-population",
   "metadata": {},
   "source": [
    "## Background\n",
    "CRISPR arrays exist in most archaeal species and many bacterial species. The arrays are encoded in chromosomal DNA as direct repeats (DR) that flank sequences (spacers) that provide immunity against invading plasmids, viral sequences, and other foreign DNA .. and RNA. These arrays can be short, comprising 1-2 spacers, or they can contain hundreds of spacer elements. The arrays are punctuated by these DR sequences. The promoter element seems to be upstream of the array, and initiates transcription of the associated RNA strand that is then cleaved and those spacers are loaded into protein machines that provide specific targeting of the associated invader should it be found. That targeting causes cleavage of the foreign DNA in known cases.\n",
    "\n",
    "There are three main classes of CRISPR systems; these are type I,II and III, and are established based on the associated protein machines - actually the specific genes that encode for them. In the archaea, only type I and III have been found so far, while in bacteria, all three types have been seen. One of those types, type II, exists in a few species, including _Streptococcus pyogenes_, which encodes for a type II system. The specific gene that is used to classify this system is named Cas9.\n",
    "\n",
    "We see cases where Cas genes, and sometimes CRISPR arrays, exist in a viral sequence. This, of course, opens many questions about the role of the CRISPR system with respect to viruses - why would a virus carry an immune defense that targets viral sequence? The existence of these arrays suggests that the CRISPR system itself might be mobile. How does the system get established, and how does the surrounding transcription machinery get encoded on the chromosome?\n",
    "\n",
    "For this assignment, we will work to find the promoter motif. We would expect that this promoter motif and an associated B recognition element (BRE) would be present in archaeal species. What is that motif, and where is it located relative to transcription start?\n",
    "\n",
    "We know little about the sequence specifics of the promoter element, though it seems to be about 13 nucleotides in length. In Pyrobaculum species, genes are packed quite tightly, so we can look upstream of the array by about 50 bases to see if a common motif sequence might exist. We need to assume that some mutations can be tolerated in that sequence.\n",
    "\n",
    "## Null model\n",
    "In prior years, we included an option that scrambled the input sequences before running to find a best score. This would effectively eliminate any signal that might be present in the data, while preserving the overall composition of the sequences. We would then compare the best score achieved without scrambling to the score achieved with scrambling. The information that remains in a scrambled set of sequences is only the base level composition of the sequences. We now use Relative Entropy to do this comparison of our experimental model(P) to our null(Q):\n",
    "\n",
    "$Q=Pr_{s\\in\\lbrace{ACTG}\\rbrace}(s)=\\frac{count(s)}{N}$\n",
    "\n",
    "We can then use relative entropy to find a score relative to the base level composition by:\n",
    "\n",
    "$REscore = \\sum_{i=0}^{cols}\\sum^{j \\in ACGT}P_{i}(j)\\log_{2}(\\frac{P_{i}(j)}{Q(j)} )$\n",
    "\n",
    "## Assignment\n",
    "__Write a BME205 style program (.py file)__ that is based on the Randomized Motif Search presented in class/videos and is described in Chapter 2 of the text. Your program should accept a fasta file that contains multiple fasta records as input (STDIN), and your program will then output (STDOUT) the consensus sequence and associated profile score. The score will be the sum of encoding costs (entropies) across each position in the final profile. We will use pseudocounts for this assignment, as described in the text (p. 86-89), and we will need an option to provide the number of pseudocounts (-p).\n",
    "\n",
    "__Produce a notebook that should be fully runable for use during inspections.__\n",
    "\n",
    "I find it easier to accumulate counts rather than calculating a probability based profile. This will also make it a bit easier to deal with pseudocounts when scoring your count-based profile.\n",
    "\n",
    "$score = \\sum_{i=0}^{cols}\\sum^{j \\in ACGT}P_{i}(j)\\log_{2}(\\frac{P_{i}(j)}{Q(j)} )$\n",
    "\n",
    "Randomized Motif Search (p. 93) can be easily trapped in local minima, so we will need to iterate some number of times to find a trajectory that produces the best results. This will need to be an option (-i) that establishes the iteration number.\n",
    "\n",
    "We also don't know the appropriate motif length that we are looking for, so we need an option (-k) to specify this.\n",
    "\n",
    "The full command line string should then look something like this:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "induced-mistake",
   "metadata": {},
   "outputs": [],
   "source": [
    "randomizedMotifSearch.py -i=100000 -p=1 -k=13 <somefile.fa >someOutputFile.fa"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fifteen-three",
   "metadata": {},
   "source": [
    "your program will need to use standard OO style, include docstrings and be well commented. It must run! \n",
    "\n",
    "I will provide a few fasta files that present the upstream 50 bases from species of the Pyrobaculum group. It is possible that the promotor+BRE that we are looking for is conserved across these species, maybe not perfectly though. Providing more data to a single run may be useful.\n",
    "\n",
    "I do have expression data for these species, so we know that not all of these arrays are functional - at least under the conditions when I grew them.\n",
    "\n",
    "Your program should output the final score achieved, along with the associated consensus motif.\n",
    "\n",
    "## Extra Credit\n",
    "The following optional features would be useful and considered for a combined extra 5 points:\n",
    "\n",
    "1) -g Use Gibbs sampling to find the optimal consensus motif.\n",
    "\n",
    "2) -m Print the specific motif and the name of the contributing sequence for each of the participating fasta records.\n",
    "\n",
    "Each of these options should only be true when the flag is specified by the user (ie. you shouldn't have to explicitly state True/False on the commandline).\n",
    "\n",
    "Submit your working program to Canvas.\n",
    "\n",
    "## Notes:\n",
    "\n",
    "1) It is likely that we will need to amend this assignment as we work with it.\n",
    "\n",
    "2) Sept-2021 eliminated the need for scrambling sequences by using relative entropy"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "equipped-remark",
   "metadata": {},
   "source": [
    "## Inspection intro\n",
    "\n",
    "- What data structure did you use to hold your sequence motifs (DNA)? \n",
    "I used pandas dataframes.\n",
    "- Is your profile  structure based on counts or frequencies? \n",
    "My profile structure is based on counts.\n",
    "- Where did you implement iterations (mostly why did you choose this)?\n",
    "I implimented iterations into a separate function, which calls the randomizedmotifsearch function. I figured it was easier and cleaner to have the RMS function be as close to the text as possible, and simply use something else to call it.\n",
    "- If you implemented Gibbs Sampling, how did you implement recomputing of the profile and motif?\n",
    "\n",
    "## Inspection results\n",
    "\n",
    "Seeing my team members implementations, there are some similarities. I think mine's the cleanest method, though. Trevor and Konstantinos had fairly cluttered notebooks, while Dodi's was pretty concise. I ended up liking Trevor's method of calculating the null model counts so I went and took it, and made it more concise by creating it on initialization instead of having a separate function for it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "aa238faa",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import math\n",
    "\n",
    "class FastAreader:\n",
    "    def __init__(self, fname=''):\n",
    "        '''contructor: saves attribute fname '''\n",
    "\n",
    "        self.fname = fname\n",
    "        self.fileH = None\n",
    "\n",
    "    def doOpen(self):\n",
    "        if self.fname == '':\n",
    "            return sys.stdin\n",
    "        else:\n",
    "            return open(self.fname)\n",
    "\n",
    "    def readFasta(self):\n",
    "\n",
    "        header = ''\n",
    "        sequence = ''\n",
    "\n",
    "        with self.doOpen() as self.fileH:\n",
    "\n",
    "            header = ''\n",
    "            sequence = ''\n",
    "\n",
    "            # skip to first fasta header\n",
    "            line = self.fileH.readline()\n",
    "            while not line.startswith('>'):\n",
    "                line = self.fileH.readline()\n",
    "            header = line[1:].rstrip()\n",
    "\n",
    "            for line in self.fileH:\n",
    "                if line.startswith('>'):\n",
    "                    yield header, sequence\n",
    "                    header = line[1:].rstrip()\n",
    "                    sequence = ''\n",
    "                else:\n",
    "                    sequence += ''.join(line.rstrip().split()).upper()\n",
    "\n",
    "        yield header, sequence\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "553244a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "class CommandLine():\n",
    "    '''\n",
    "    Handle the command line, usage and help requests.\n",
    "\n",
    "    CommandLine uses argparse, now standard in 2.7 and beyond. \n",
    "    it implements a standard command line argument parser with various argument options,\n",
    "    a standard usage and help, and an error termination mechanism do-usage_and_die.\n",
    "\n",
    "    attributes:\n",
    "    all arguments received from the commandline using .add_argument will be\n",
    "    avalable within the .args attribute of object instantiated from CommandLine.\n",
    "    For example, if myCommandLine is an object of the class, and requiredbool was\n",
    "    set as an option using add_argument, then myCommandLine.args.requiredbool will\n",
    "    name that option.\n",
    "\n",
    "    '''\n",
    "\n",
    "    def __init__(self, inOpts=None):\n",
    "        '''\n",
    "        CommandLine constructor.\n",
    "        Implements a parser to interpret the command line argv string using argparse.\n",
    "        '''\n",
    "\n",
    "        import argparse\n",
    "        self.parser = argparse.ArgumentParser(\n",
    "            description='Program prolog - a brief description of what this thing does',\n",
    "            epilog='Program epilog - some other stuff you feel compelled to say',\n",
    "            add_help=True,  # default is True\n",
    "            prefix_chars='-',\n",
    "            usage='%(prog)s [options] -option1[default] <input >output'\n",
    "        )\n",
    "\n",
    "        self.parser.add_argument('-k', '--motif_length', type=int, default=15, action='store',\n",
    "                                 help='kMer size')\n",
    "        self.parser.add_argument('-p', '--pseudo_count', type=int, default=1, action='store',\n",
    "                                 help='pseudo count value')\n",
    "        self.parser.add_argument('-i', '--iterations', type=int, default=1000, action='store',\n",
    "                                 help='times to run RMS')\n",
    "\n",
    "\n",
    "        self.parser.add_argument('-v', '--version', action='version', version='%(prog)s 0.1')\n",
    "        if inOpts is None:\n",
    "            self.args = self.parser.parse_args()\n",
    "        else:\n",
    "            self.args = self.parser.parse_args(inOpts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "id": "aae631b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "class Motif():\n",
    "    \n",
    "    def __init__(self, DNA, k, pseudo=1, iterations=1000):\n",
    "        # initialize variables\n",
    "        self.bases = ['A', 'C', 'G', 'T']\n",
    "        self.DNA = DNA\n",
    "        self.k = k\n",
    "        self.t = t\n",
    "        self.pseudo = pseudo\n",
    "        self.iterations = 1000\n",
    "        \n",
    "        # initialize null model using concantenated DNA sequences\n",
    "        self.string =''.join(self.DNA)\n",
    "        self.null = {base: (self.pseudo + self.string.count(base))/(len(self.string)+(self.pseudo*4)) for base in self.bases}\n",
    "    \n",
    "    def rand_motifs(self):\n",
    "        # use RNG to create a bunch of random motifs\n",
    "        motifs = []\n",
    "        for seq in self.DNA:\n",
    "            start_i = np.random.randint(0, len(seq)-self.k+1)\n",
    "            end_i = start_i + self.k\n",
    "            \n",
    "            motifs.append(seq[start_i:end_i])  \n",
    "        return motifs\n",
    "\n",
    "\n",
    "    def Score(self, profile):\n",
    "        # Find relative entropy score using profile table\n",
    "        score = 0\n",
    "        for x in range(self.k):\n",
    "            for base in self.bases: \n",
    "                prob_x = profile.loc[base, x]\n",
    "                score += prob_x * np.log2( prob_x/self.null[base] )\n",
    "        return score\n",
    " \n",
    "    \n",
    "    def Count(self, motifs):\n",
    "        # construct count table that's already filled with pseudo-count\n",
    "        count_table = pd.DataFrame(    np.full((4,self.k), self.pseudo)   , index=self.bases    )\n",
    "        \n",
    "        # iterate over every motif and return a table with counts of bases\n",
    "        for motif in motifs:\n",
    "            for index, base in enumerate(motif): #bases and index correspond to rows and columns\n",
    "                count_table.loc[base, index] += 1\n",
    "        \n",
    "        return count_table\n",
    "    def Profile(self, motifs):\n",
    "        # use count table to calculate profile probabilities using base count and sum\n",
    "        count_table = Count(motifs)\n",
    "        \n",
    "        profile_table = pd.DataFrame(    np.full((4,self.k), self.pseudo)   , index=self.bases     )\n",
    "        \n",
    "        for x in range(self.k):\n",
    "            total = sum(count_table[x])\n",
    "            for base in self.bases:#           count / total\n",
    "                profile_table.loc[base, x] = count_table.loc[base, x] / total\n",
    "        return profile_table\n",
    "\n",
    "    \n",
    "    def Consensus(self, profile):\n",
    "        # iterates over a profile and returns the consensus string according to\n",
    "        # the max value and corresponding row (base) of each column\n",
    "        consensus_string = ''\n",
    "        for x in range(self.k):\n",
    "            consensus_string += profile[x].idxmax()\n",
    "            \n",
    "        return consensus_string\n",
    "    \n",
    "    def motifs_from_profile(self, DNA, profile):\n",
    "        # uses profile table to find the kmer with the best cumulative probability\n",
    "        # in each sequence\n",
    "        motifs = []\n",
    "        for seq in DNA:\n",
    "            best_prob = -1\n",
    "            for x in range(len(seq)-self.k+1):\n",
    "                kmer = seq[x:x+k]\n",
    "                prob = 1\n",
    "                for index, base in enumerate(kmer):\n",
    "                    prob *= profile.loc[base, index]\n",
    "                if prob > best_prob:\n",
    "                    best_prob = prob\n",
    "                    best_motif = kmer\n",
    "            motifs.append(best_motif)\n",
    "        return motifs\n",
    "     \n",
    "    def RandomizedMotifSearch(self):\n",
    "        # conduct a depth-first search from randomly selected kmers\n",
    "        best_motifs = self.rand_motifs()\n",
    "        best_score = self.Score( self.Profile(best_motifs)  )\n",
    "\n",
    "        while 1:\n",
    "            profile = self.Profile(best_motifs)\n",
    "            motifs = self.motifs_from_profile(self.DNA, profile)\n",
    "            \n",
    "            score = self.Score( self.Profile(motifs) )\n",
    "            if score > best_score:\n",
    "                best_motifs = motifs\n",
    "                best_score = score\n",
    "\n",
    "            else:\n",
    "                return [best_motifs, best_score]\n",
    "    \n",
    "    def complete_search(self):\n",
    "        # repeat randomized motif search a number of times and return the\n",
    "        # consensus string with the best score\n",
    "        final_list = []\n",
    "        for x in range(self.iterations):\n",
    "            motifs_candidate = self.RandomizedMotifSearch()\n",
    "            final_list.append(motifs_candidate)\n",
    "        final_motifs = sorted(final_list, key=lambda x: x[1], reverse=True)\n",
    "        motifs, score = final_motifs[0]\n",
    "        profile = self.Profile(motifs)\n",
    "        return self.Consensus(profile), score\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "id": "d8856a63",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('GTGTACAG', 2.709372622306024)\n"
     ]
    }
   ],
   "source": [
    "\n",
    "k = 8\n",
    "t = 5\n",
    "DNA = [\n",
    "    'CGCCCCTCTCGGGGGTGTTCAGTAAACGGCCA',\n",
    "    'GGGCGAGGTATGTGTAAGTGCCAAGGTGCCAG',\n",
    "    'TAGTACCGAGACCGAAAGAAGTATACAGGCGT',\n",
    "    'TAGATCAAGTTTCAGGTGCACGTCGGTGAACC',\n",
    "    'AATCCACCAGCTCCACGTGCAATGTTGGCCTA'\n",
    "]\n",
    "\n",
    "cheese = Motif(DNA, k)\n",
    "\n",
    "THE_MOTIF = cheese.complete_search()\n",
    "print(THE_MOTIF)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "909a054c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  },
  "nteract": {
   "version": "0.28.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
